File: static/js/shortterm_pax_demand.js
/* Short Term PAX Demand - 3-day passenger flow outlook */

let STPAX_DATA = null;
let STPAX_CHART = null;

const STPAX_SERIES = [
  { key: 'checkin', label: 'Check-in (Inflow)', color: '#0ea5e9', axis: 'y' },
  { key: 'security', label: 'Security Queueing', color: '#8b5cf6', axis: 'y' },
  { key: 'cbp', label: 'US Preclearance (CBP)', color: '#ef4444', axis: 'y' },
  { key: 'lounge', label: 'Lounge/Retail Occupancy', color: '#f59e0b', axis: 'y1', dash: [5, 5] },
  { key: 'boarding', label: 'Boarding Gate Pulse', color: '#ec4899', axis: 'y' },
  { key: 'immigration', label: 'Immigration', color: '#10b981', axis: 'y' },
  { key: 'baggage', label: 'Baggage Reclaim Surge', color: '#3b82f6', axis: 'y' },
];

const STPAX_ACCENT = {
  teal: '#0ea5e9',
  purple: '#8b5cf6',
  crit: '#ef4444',
  warn: '#f59e0b',
  info: '#3b82f6',
  ok: '#10b981',
};

const STPAX_TOUCHPOINT_ICONS = {
  check: '<path d="M3 7h18"/><path d="M5 7l2-4h10l2 4"/><path d="M6 11h12"/><path d="M8 15h8"/><path d="M10 19h4"/>',
  security: '<path d="M12 22s8-4 8-10V5l-8-3-8 3v7c0 6 8 10 8 10z"/><path d="m9 12 2 2 4-5"/>',
  cbp: '<path d="M4 4h16v16H4z"/><path d="M8 8h8"/><path d="M8 12h8"/><path d="M8 16h5"/>',
  lounge: '<path d="M4 11h16"/><path d="M5 11l1 8h12l1-8"/><path d="M8 11V7a4 4 0 0 1 8 0v4"/>',
  boarding: '<path d="M22 2 11 13"/><path d="M22 2 15 22l-4-9-9-4 20-7z"/>',
  immigration: '<path d="M16 21v-2a4 4 0 0 0-8 0v2"/><circle cx="12" cy="7" r="4"/><path d="M4 3h16v18H4z"/>',
  baggage: '<rect x="5" y="7" width="14" height="13" rx="2"/><path d="M9 7V5a2 2 0 0 1 2-2h2a2 2 0 0 1 2 2v2"/><path d="M9 11v5"/><path d="M15 11v5"/>',
  default: '<path d="M3 3v18h18"/><path d="m7 15 4-4 3 3 5-7"/>',
};

function stpaxFormat(value) {
  return Number(value || 0).toLocaleString();
}

function stpaxEsc(value) {
  return String(value ?? '').replace(/[&<>"']/g, ch => ({
    '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;',
  }[ch]));
}

function stpaxSetText(id, value) {
  const el = document.getElementById(id);
  if (el) el.textContent = value;
}

function stpaxDayShort(label) {
  return String(label || '').split(' ').slice(1).join(' ');
}

function stpaxIconFor(label) {
  const value = String(label || '').toLowerCase();
  if (value.includes('check')) return STPAX_TOUCHPOINT_ICONS.check;
  if (value.includes('security')) return STPAX_TOUCHPOINT_ICONS.security;
  if (value.includes('cbp') || value.includes('preclearance')) return STPAX_TOUCHPOINT_ICONS.cbp;
  if (value.includes('lounge') || value.includes('retail')) return STPAX_TOUCHPOINT_ICONS.lounge;
  if (value.includes('boarding') || value.includes('gate')) return STPAX_TOUCHPOINT_ICONS.boarding;
  if (value.includes('immigration')) return STPAX_TOUCHPOINT_ICONS.immigration;
  if (value.includes('baggage') || value.includes('reclaim')) return STPAX_TOUCHPOINT_ICONS.baggage;
  return STPAX_TOUCHPOINT_ICONS.default;
}

function stpaxRenderInsights(data) {
  const grid = document.getElementById('stpax-insight-grid');
  if (!grid) return;

  grid.innerHTML = (data.insights || []).map(item => {
    const accent = STPAX_ACCENT[item.accent] || STPAX_ACCENT.info;
    const icon = stpaxIconFor(item.label);
    return `
      <article class="stpax-insight-card" style="--stpax-accent:${accent};--pax-accent:${accent}">
        <div class="stpax-card-head">
          <div class="pax-icon-bubble" aria-hidden="true">
            <svg width="22" height="22" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.8" stroke-linecap="round" stroke-linejoin="round">${icon}</svg>
          </div>
          <div class="stpax-card-name">${stpaxEsc(item.label)}</div>
        </div>
        <div class="stpax-card-value">${stpaxFormat(item.peak)}</div>
        <div class="stpax-card-sub">${stpaxEsc(item.metric_label || 'Peak Pax / 15-min')}</div>
        <div class="stpax-card-row">
          <span>Peak Window</span>
          <strong>${stpaxEsc(item.peak_time || '--')}</strong>
        </div>
        <div class="stpax-card-row">
          <span>${stpaxEsc(item.secondary_label || '')}</span>
          <strong>${stpaxEsc(item.secondary_value || '--')}</strong>
        </div>
      </article>
    `;
  }).join('');
}

function stpaxRenderChart(payload) {
  const canvas = document.getElementById('stpax-outlook-chart');
  if (!canvas || !window.Chart) return;

  if (STPAX_CHART) {
    STPAX_CHART.destroy();
    STPAX_CHART = null;
  }

  const series = payload.series || {};
  const labels = series.labels || [];
  const dayLabels = series.day_labels || [];

  STPAX_CHART = new Chart(canvas.getContext('2d'), {
    type: 'line',
    data: {
      labels,
      datasets: STPAX_SERIES.map(item => ({
        label: item.label,
        data: series[item.key] || [],
        borderColor: item.color,
        backgroundColor: `${item.color}22`,
        fill: true,
        tension: 0.35,
        borderWidth: 2,
        pointRadius: 0,
        hitRadius: 10,
        hoverRadius: 3,
        yAxisID: item.axis,
        borderDash: item.dash || [],
      })),
    },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      interaction: { mode: 'index', intersect: false },
      elements: { point: { radius: 0 } },
      plugins: {
        legend: {
          position: 'top',
          labels: {
            usePointStyle: true,
            pointStyle: 'rectRounded',
            boxWidth: 14,
            boxHeight: 6,
            padding: 18,
          },
        },
        tooltip: {
          mode: 'index',
          intersect: false,
          callbacks: {
            title: items => {
              const idx = items[0]?.dataIndex || 0;
              return `${dayLabels[idx] || ''} ${labels[idx] || ''}`.trim();
            },
            label: ctx => `${ctx.dataset.label}: ${Math.round(ctx.parsed.y || 0).toLocaleString()}`,
          },
        },
      },
      scales: {
        x: {
          grid: { color: 'rgba(148, 163, 184, 0.11)' },
          ticks: {
            maxRotation: 0,
            autoSkip: false,
            callback: function(_value, index) {
              const label = labels[index] || '';
              const [h, m] = label.split(':').map(Number);
              if (m !== 0 || h % 3 !== 0) return '';
              const prevDay = index > 0 ? dayLabels[index - 1] : '';
              const currentDay = dayLabels[index] || '';
              const hour = String(h);
              if (currentDay && currentDay !== prevDay) return [hour, stpaxDayShort(currentDay)];
              return hour;
            },
          },
        },
        y: {
          min: 0,
          title: { display: true, text: 'Pax / 15-min' },
          grid: { color: 'rgba(148, 163, 184, 0.16)' },
          ticks: { callback: value => Number(value).toLocaleString() },
        },
        y1: {
          min: 0,
          position: 'right',
          title: { display: true, text: 'Lounge Occupancy' },
          grid: { drawOnChartArea: false },
          ticks: { callback: value => Number(value).toLocaleString() },
        },
      },
    },
  });
}

const STPAX_MONTHS = { Jan:1, Feb:2, Mar:3, Apr:4, May:5, Jun:6, Jul:7, Aug:8, Sep:9, Oct:10, Nov:11, Dec:12 };

function stpaxDayToISO(dayStr) {
  // "Fri 16 May 2025" -> "2025-05-16"
  const parts = String(dayStr || '').split(' ');
  if (parts.length < 4) return '';
  const d = parts[1].padStart(2, '0');
  const m = String(STPAX_MONTHS[parts[2]] || 0).padStart(2, '0');
  const y = parts[3];
  return `${y}-${m}-${d}`;
}

function stpaxComputeInsightsFromSeries(series) {
  const nums = key => (series[key] || []).map(v => Number(v || 0));
  const peakVal = values => Math.round(Math.max(0, ...values));
  const activeAvg = values => {
    const active = values.filter(v => v > 0);
    return active.length ? Math.round(active.reduce((a, b) => a + b, 0) / active.length) : 0;
  };
  const pct = (part, total) => total ? Math.round((part / total) * 100) : null;
  const peakTime = key => {
    const vals = nums(key);
    if (!vals.length) return '--';
    const idx = vals.indexOf(Math.max(...vals));
    return (series.labels || [])[idx] || '--';
  };
  const fmt = v => Number(v || 0).toLocaleString();

  const checkin = nums('checkin'), security = nums('security'), cbp = nums('cbp');
  const lounge = nums('lounge'), boarding = nums('boarding');
  const immigration = nums('immigration'), baggage = nums('baggage');
  const maxCheckin = peakVal(checkin), maxSecurity = peakVal(security), maxCbp = peakVal(cbp);
  const maxLounge = peakVal(lounge), maxBoarding = peakVal(boarding);
  const maxImmigration = peakVal(immigration), maxBaggage = peakVal(baggage);
  const avgSecurity = activeAvg(security);

  return [
    { key: 'checkin',    label: 'Check-In',         accent: 'teal',   peak: maxCheckin,    metric_label: 'Peak Pax / 15-min',    peak_time: peakTime('checkin'),    secondary_label: 'Avg (all day)',        secondary_value: `${fmt(activeAvg(checkin))} pax` },
    { key: 'security',   label: 'Security',          accent: 'purple', peak: maxSecurity,   metric_label: 'Peak Pax / 15-min',    peak_time: peakTime('security'),   secondary_label: 'Pressure Index',      secondary_value: avgSecurity ? `${(maxSecurity / avgSecurity).toFixed(1)}x avg` : '--' },
    { key: 'cbp',        label: 'CBP Preclearance',  accent: 'crit',   peak: maxCbp,        metric_label: 'Peak Pax / 15-min',    peak_time: peakTime('cbp'),        secondary_label: '% of Security',       secondary_value: pct(maxCbp, maxSecurity) != null ? `${pct(maxCbp, maxSecurity)}% of security peak` : '--' },
    { key: 'lounge',     label: 'Lounge / Retail',   accent: 'warn',   peak: maxLounge,     metric_label: 'Peak Concurrent Pax',  peak_time: peakTime('lounge'),     secondary_label: 'Avg Occupancy',       secondary_value: `${fmt(activeAvg(lounge))} pax` },
    { key: 'boarding',   label: 'Boarding Gates',    accent: 'info',   peak: maxBoarding,   metric_label: 'Peak Pax / 15-min',    peak_time: peakTime('boarding'),   secondary_label: 'Gate Pressure',       secondary_value: pct(maxBoarding, maxCheckin) != null ? `${pct(maxBoarding, maxCheckin)}% of check-in` : '--' },
    { key: 'immigration',label: 'Immigration',       accent: 'ok',     peak: maxImmigration,metric_label: 'Peak Pax / 15-min',    peak_time: peakTime('immigration'),secondary_label: 'Intl. Arrival Share', secondary_value: pct(maxImmigration, maxBaggage) != null ? `${pct(maxImmigration, maxBaggage)}% of arrivals` : '--' },
  ];
}

function stpaxFilterDay(payload, isoDate) {
  const daysData = payload.series?.days_data || [];
  const day = daysData.find(d => stpaxDayToISO(d.date) === isoDate);
  if (!day) return payload;

  const filteredSeries = {
    labels: day.labels || [],
    day_labels: (day.labels || []).map(() => day.date),
    days_data: [day],
  };
  for (const key of ['checkin', 'security', 'cbp', 'lounge', 'boarding', 'immigration', 'baggage']) {
    filteredSeries[key] = day[key] || [];
  }

  const parts = String(day.date || '').split(' ');
  const dayShort = parts.slice(1, 3).join(' ');
  return {
    ...payload,
    summary: { ...payload.summary, title: `${dayShort} Passenger Flow`, subtitle: 'passenger flow for the selected date' },
    series: filteredSeries,
    insights: stpaxComputeInsightsFromSeries(filteredSeries),
  };
}

function stpaxRender(payload) {
  stpaxSetText('stpax-title', payload.summary?.title || '3-Day Strategic Outlook');
  stpaxSetText('stpax-subtitle', payload.summary?.subtitle || '');
  stpaxRenderChart(payload);
  stpaxRenderInsights(payload);
}

async function initShortTermPaxDemand(options = {}) {
  const panel = document.querySelector('.stpax-outlook-panel');
  if (!panel) return;

  try {
    if (!STPAX_DATA || options.force) {
      const res = await fetch('/api/short-term/pax-demand/strategic-outlook');
      if (!res.ok) throw new Error(`Short-term PAX API failed: ${res.status}`);
      STPAX_DATA = await res.json();
    }
    const renderPayload = options.date ? stpaxFilterDay(STPAX_DATA, options.date) : STPAX_DATA;
    stpaxRender(renderPayload);
  } catch (err) {
    panel.innerHTML = '<div class="empty-state">Unable to load Short Term PAX Demand outlook.</div>';
    console.error(err);
  }
}

window.initShortTermPaxDemand = initShortTermPaxDemand;

File: short_term_pax_demand.py
import json
import os
from collections import defaultdict
from datetime import datetime, timedelta


PAX_3DAY_FILE = 'pax_3day_cache.json'
PAX_PROFILE_FILE = 'short term PAX.xlsx'
TOUCHPOINTS = ('checkin', 'security', 'cbp', 'lounge', 'boarding', 'immigration', 'baggage')


def _data_path(base_dir, filename):
    return os.path.join(base_dir, 'data', filename)


def _load_json(base_dir, filename):
    with open(_data_path(base_dir, filename), encoding='utf-8') as f:
        return json.load(f)


def _numbers(values):
    out = []
    for value in values or []:
        try:
            out.append(float(value or 0))
        except (TypeError, ValueError):
            out.append(0.0)
    return out


def _peak_idx(values):
    if not values:
        return 0
    return max(range(len(values)), key=lambda idx: values[idx])


def _active_avg(values):
    active = [v for v in values if v > 0]
    return round(sum(active) / len(active)) if active else 0


def _pct(part, total):
    if not total:
        return None
    return round((part / total) * 100)


def _load_pax_workbook_rows(base_dir):
    path = _data_path(base_dir, PAX_PROFILE_FILE)
    if not os.path.exists(path):
        return {'dates': [], 'rows_by_date': {}}

    rows_by_date = defaultdict(list)
    try:
        import openpyxl
        wb = openpyxl.load_workbook(path, data_only=True, read_only=True)
        ws = wb.active
        headers = [str(v).strip() if v is not None else '' for v in next(ws.iter_rows(min_row=1, max_row=1, values_only=True))]
        for vals in ws.iter_rows(min_row=2, values_only=True):
            row = dict(zip(headers, vals))
            ts = row.get('Timestamp')
            if isinstance(ts, datetime):
                rows_by_date[ts.date()].append(row)
    except Exception:
        return {'dates': [], 'rows_by_date': {}}

    return {'dates': sorted(rows_by_date.keys()), 'rows_by_date': dict(rows_by_date)}


def _source_date_for_target(target_dt, source_dates):
    target_date = target_dt.date()
    if target_date in source_dates:
        return target_date

    offset = (target_date - datetime.now().date()).days
    if 0 <= offset < len(source_dates):
        return source_dates[offset]
    return None


def _touchpoint_value(row, key):
    total = 0.0
    suffix = f"_{key}"
    for col, value in row.items():
        if str(col or '').strip().lower().endswith(suffix):
            try:
                total += float(value or 0)
            except (TypeError, ValueError):
                pass
    return total


def _empty_series():
    data = {'labels': [], 'day_labels': [], 'days_data': []}
    for key in TOUCHPOINTS:
        data[key] = []
    return data


def _build_series_from_workbook(base_dir):
    profile = _load_pax_workbook_rows(base_dir)
    source_dates = profile.get('dates', [])
    rows_by_date = profile.get('rows_by_date', {})
    if not source_dates:
        return None

    data = _empty_series()
    today = datetime.now()
    for offset in (1, 2, 3):
        target_dt = today + timedelta(days=offset)
        source_date = _source_date_for_target(target_dt, source_dates)
        if source_date is None:
            continue

        day = {
            'date': target_dt.strftime('%a %d %b %Y'),
            'labels': [],
        }
        for key in TOUCHPOINTS:
            day[key] = []

        rows = sorted(rows_by_date.get(source_date, []), key=lambda r: r.get('Timestamp'))
        for row in rows:
            ts = row.get('Timestamp')
            if not isinstance(ts, datetime):
                continue

            label = ts.strftime('%H:%M')
            day_label = target_dt.strftime('%a %d %b')
            data['labels'].append(label)
            data['day_labels'].append(day_label)
            day['labels'].append(label)
            for key in TOUCHPOINTS:
                value = _touchpoint_value(row, key)
                data[key].append(value)
                day[key].append(value)

        data['days_data'].append(day)

    return data if data['labels'] else None


def _build_series_from_cache(base_dir):
    source = _load_json(base_dir, PAX_3DAY_FILE)
    data = {
        'labels': source.get('labels', []),
        'day_labels': source.get('day_labels', []),
        'days_data': source.get('days_data', []),
    }
    for key in TOUCHPOINTS:
        data[key] = _numbers(source.get(key))
    return data


def _peak_time_label(data, values):
    idx = _peak_idx(values)
    time_label = (data.get('labels') or ['--'])[idx] if idx < len(data.get('labels') or []) else '--'
    day_label = (data.get('day_labels') or [''])[idx] if idx < len(data.get('day_labels') or []) else ''
    day_short = ' '.join(str(day_label).split(' ')[1:])
    return f"{day_short} · {time_label}" if day_short else time_label


def _insight(data, key, label, accent, secondary_label, secondary_value, metric_label='Peak Pax / 15-min'):
    values = _numbers(data.get(key))
    return {
        'key': key,
        'label': label,
        'accent': accent,
        'peak': round(max(values) if values else 0),
        'metric_label': metric_label,
        'peak_time': _peak_time_label(data, values),
        'secondary_label': secondary_label,
        'secondary_value': secondary_value,
    }


def build_short_term_pax_outlook(base_dir):
    data = _build_series_from_workbook(base_dir) or _build_series_from_cache(base_dir)

    max_checkin = max(data['checkin'] or [0])
    max_security = max(data['security'] or [0])
    max_cbp = max(data['cbp'] or [0])
    max_lounge = max(data['lounge'] or [0])
    max_boarding = max(data['boarding'] or [0])
    max_immigration = max(data['immigration'] or [0])
    max_baggage = max(data['baggage'] or [0])
    avg_security = _active_avg(data['security'])

    insights = [
        _insight(data, 'checkin', 'Check-In', 'teal', 'Avg (all day)', f"{_active_avg(data['checkin']):,} pax"),
        _insight(data, 'security', 'Security', 'purple', 'Pressure Index', f"{(max_security / avg_security):.1f}x avg" if avg_security else '--'),
        _insight(data, 'cbp', 'CBP Preclearance', 'crit', '% of Security', f"{_pct(max_cbp, max_security) or '--'}% of security peak"),
        _insight(data, 'lounge', 'Lounge / Retail', 'warn', 'Avg Occupancy', f"{_active_avg(data['lounge']):,} pax", 'Peak Concurrent Pax'),
        _insight(data, 'boarding', 'Boarding Gates', 'info', 'Gate Pressure', f"{_pct(max_boarding, max_checkin) or '--'}% of check-in"),
        _insight(data, 'immigration', 'Immigration', 'ok', 'Intl. Arrival Share', f"{_pct(max_immigration, max_baggage) or '--'}% of arrivals"),
    ]

    return {
        'summary': {
            'title': '3-Day Strategic Outlook',
            'subtitle': 'passenger numbers engineered based on actual next 3-day flight schedules',
            'interval': '15-min',
            'lounge_peak': round(max_lounge),
        },
        'series': data,
        'insights': insights,
    }
